Sources/Main.cpp

```diff
@@ -1268,5 +1280,126 @@ void __fastcall TX584Form::AboutItemClick(TObject *Sender)
 {
     AboutForm->ShowModal();
 }
+
+//---------------------------------------------------------------------------
+//                           *** БУФЕР ОБМЕНА ***
+//---------------------------------------------------------------------------
+
+int TX584Form::GetClipboardBufferSize()
+{
+    int result = 0;
+
+    result += 2; // количество элементов в буфере
+    for (int i = 0; i < ClipboardSize; i++) {
+		result += 2 + 1 + CMClipboard[i].Length() * 2;
+    }
+
+    return result;
+}
+//---------------------------------------------------------------------------
+
+void TX584Form::PutIntoClipboard()
+{
+    HGLOBAL hBuffer = NULL;
+    unsigned char *Buffer, *p;
+    int BufferSize;
+
+    Clipboard()->Open();
+    Clipboard()->Clear();
+    BufferSize = GetClipboardBufferSize();
+
+    hBuffer = GlobalAlloc(GMEM_MOVEABLE, BufferSize);
+    Buffer = (unsigned char *)GlobalLock(hBuffer);
+    p = Buffer;
+
+    // записать количество элементов в буфере обмена
+    *p++ = (ClipboardSize & 0xFF);
+    *p++ = ((ClipboardSize >> 8) & 0xFF);
+    // записать каждый элемент
+    for (int i = 0; i < ClipboardSize; i++) {
+        // записать микроинструкцию
+        *p++ = (MIClipboard[i] & 0xFF);
+        *p++ = ((MIClipboard[i] >> 8) & 0xFF);
+        // записать длину комментария
+        *p++ = (CMClipboard[i].Length() & 0xFF);
+        // записать комментарий
+		for (int j = 1; j <= CMClipboard[i].Length(); j++) {
+			WideChar Ch = CMClipboard[i][j];
+			*p++ = ((int)Ch)&0xFF;
+			*p++ = (((int)Ch) >> 8) & 0xFF;
+        }
+    }
+    GlobalUnlock(hBuffer);
+    Clipboard()->SetAsHandle(ClipboardFormat, (unsigned)hBuffer);
+    Clipboard()->Close();
+}
+//---------------------------------------------------------------------------
+
+void TX584Form::GetFromClipboard()
+{
+    HGLOBAL hBuffer;
+    unsigned char *Buffer, *p;
+
+    Clipboard()->Open();
+    if (!Clipboard()->HasFormat(ClipboardFormat)) {
+        ClipboardSize = 0;
+        return;
+    }
+
+    hBuffer = (HGLOBAL)Clipboard()->GetAsHandle(ClipboardFormat);
+    Buffer = (unsigned char *)GlobalLock(hBuffer);
+
+    p = Buffer;
+    // считать количество элементов в буфере обмена
+    ClipboardSize = 0;
+    ClipboardSize = *p++;
+    ClipboardSize |= (*p++) << 8;
+
+    for (int i = 0; i < ClipboardSize; i++) {
+        // считать код микроинструкции
+        int Opcode = 0;
+        Opcode = *p++;
+        Opcode |= (*p++) << 8;
+        MIClipboard[i] = Opcode;
+
+        // считать длину комментария
+        int CommentSize = *p++;
+        // считать сам комментарий
+		CMClipboard[i] = UnicodeString();
+		for (int j = 0; j < CommentSize; j++) {
+			WideChar Ch;
+			Ch = *p++;
+			Ch |= (*p++) << 8;
+			CMClipboard[i] += Ch;
+        }
+    }
+    GlobalUnlock(hBuffer);
+    Clipboard()->Close();
+}
+//---------------------------------------------------------------------------
+
+void TX584Form::OnClipboard(TWMNoParams &x)
+{
+    if (Clipboard()->HasFormat(ClipboardFormat)) {
```

Comment:
This can be simplified to the following code:

```cpp
bool ClipboardHasData = Clipboard()->HasFormat(ClipboardFormat);
PasteItem->Enabled = ClipboardHasData;
PasteToolButton->Enabled = ClipboardHasData;
```